apollo-federation/src/connectors/runtime/cache.rs

```diff
@@ -0,0 +1,794 @@
+                            surrogate_key_data,
+                        },
+                        cache_components.clone(), // Clone for each batch item
+                    ));
+                }
+            }
+        }
+    }
+
+    // Consolidate root fields if any exist
+    if !root_field_requests.is_empty() {
+        let mut consolidated_components = CacheKeyComponents {
+            subgraph_name: subgraph_name.to_string(),
+            methods: Vec::new(),
+            uris: Vec::new(),
+            headers: Vec::new(),
+            bodies: Vec::new(),
+        };
+
+        // Consolidate data from all root field requests
+        for transport_request in &root_field_requests {
+            let individual_components = extract_cache_components(subgraph_name, transport_request);
+            consolidated_components
+                .methods
+                .extend(individual_components.methods);
+            consolidated_components
+                .uris
+                .extend(individual_components.uris);
+            consolidated_components
+                .headers
+                .extend(individual_components.headers);
+            consolidated_components
+                .bodies
+                .extend(individual_components.bodies);
+        }
+
+        // Extract debugging information from the first root field key
+        // (assuming all root fields have the same operation_type and output_type for consolidation)
+        let first_root_key = &root_field_keys.first();
+        let (operation_type, output_type, output_names, surrogate_key_data) = match first_root_key {
+            Some(ResponseKey::RootField {
+                operation_type,
+                output_type,
+                inputs,
+                ..
+            }) => {
+                let names = root_field_keys
+                    .iter()
+                    .filter_map(|key| {
+                        if let ResponseKey::RootField { name, .. } = key {
+                            Some(name.clone())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                // For RootFields, use inputs.args for surrogate key data
+                let surrogate_key_data = inputs.args.clone();
+                (
+                    *operation_type,
+                    output_type.clone(),
+                    names,
+                    surrogate_key_data,
+                )
+            }
+            _ => unreachable!("root_field_keys should only contain RootField variants"),
+        };
+
+        items.insert(
+            0,
+            (
+                CacheableItem::RootFields {
+                    operation_type,
+                    output_type,
+                    output_names,
+                    surrogate_key_data,
+                },
+                consolidated_components,
+            ),
+        );
```

Comment:
Does it really need to be first ? if so i'm wondering if we can't do this directly in the right match arm branch to avoid moving unnecessary data.